### PR TITLE
[RISCV][GISEL] Legalize  G_INSERT_SUBVECTOR

### PR DESCRIPTION
**Pull request by michaelmaitland**
This code is heavily based on the SelectionDAG lowerINSERT_SUBVECTOR code.

### CODE REVIEW

**Comment by topperc** [comment]
How does one create a G_INSERT_SUBVECTOR today? There are no callers of `MachineIRBuilder::buildInsertSubvector`

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -915,6 +921,154 @@ bool RISCVLegalizerInfo::legalizeSplatVector(MachineInstr &MI,
...
+
+  const RISCVRegisterInfo *TRI = STI.getRegisterInfo();
+  MVT LitTyMVT = getMVTForLLT(LitTy);
+  unsigned SubRegIdx, RemIdx;
+  std::tie(SubRegIdx, RemIdx) =
+      RISCVTargetLowering::decomposeSubvectorInsertExtractToSubRegs(
+          getMVTForLLT(BigTy), LitTyMVT, Idx, TRI);
+
+  RISCVII::VLMUL SubVecLMUL = RISCVTargetLowering::getLMUL(getMVTForLLT(LitTy));
+  bool IsSubVecPartReg = SubVecLMUL == RISCVII::VLMUL::LMUL_F2 ||
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:992]: Can this use the fractional flag returned from `decodeVLMUL`? Maybe we could make the same change to SelectionDAG too.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -915,6 +921,154 @@ bool RISCVLegalizerInfo::legalizeSplatVector(MachineInstr &MI,
...
+    VL = MIB.buildAdd(XLenTy, SlideupAmt, VL);
+    Inserted =
+        MIB.buildInstr(RISCV::G_VSLIDEUP_VL, {InterLitTy},
+                       {AlignedExtract, LitVec, SlideupAmt, Mask, VL, Policy});
+  }
+
+  // If required, insert this subvector back into the correct vector register.
+  // This should resolve to an INSERT_SUBREG instruction.
+  if (TypeSize::isKnownGT(BigTy.getSizeInBits(), InterLitTy.getSizeInBits()))
+    Inserted = MIB.buildInsert(BigTy, BigVec, LitVec, AlignedIdx);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1062]: InsertSubvector?

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-insert-subvector.mir [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    ; CHECK-NEXT: $v8 = COPY [[INSERT_SUBVECTOR]](<vscale x 64 x s1>)
+    ; CHECK-NEXT: PseudoRET implicit $v8
+    %0:_(<vscale x 64 x s1>) = G_IMPLICIT_DEF
+    %1:_(<vscale x 16 x s1>) = G_IMPLICIT_DEF
+    %2:_(<vscale x 64 x s1>) = G_INSERT_SUBVECTOR %0(<vscale x 64 x s1>), %1, 0
+    $v8 = COPY %2(<vscale x 64 x s1>)
+    PseudoRET implicit $v8
+...
+
+# Inserrt with zero index
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-insert-subvector.mir:256]: Insert*

**Comment by michaelmaitland** [comment]
> How does one create a G_INSERT_SUBVECTOR today? There are no callers of `MachineIRBuilder::buildInsertSubvector`

It will be needed by G_INSERT_VECTOR_ELT. This is a precursor to that patch.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-insert-subvector.mir [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+    ; CHECK-NEXT: [[DEF1:%[0-9]+]]:_(<vscale x 1 x s64>) = G_IMPLICIT_DEF
+    ; CHECK-NEXT: [[INSERT_SUBVECTOR:%[0-9]+]]:_(<vscale x 8 x s64>) = G_INSERT_SUBVECTOR [[DEF]], [[DEF1]](<vscale x 1 x s64>), 0
+    ; CHECK-NEXT: $v8 = COPY [[INSERT_SUBVECTOR]](<vscale x 8 x s64>)
+    ; CHECK-NEXT: PseudoRET implicit $v8
+    %0:_(<vscale x 8 x s64>) = G_IMPLICIT_DEF
+    %1:_(<vscale x 1 x s64>) = G_IMPLICIT_DEF
+    %2:_(<vscale x 8 x s64>) = G_INSERT_SUBVECTOR %0(<vscale x 8 x s64>), %1, 0
+    $v8 = COPY %2(<vscale x 8 x s64>)
+    PseudoRET implicit $v8
+...
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-insert-subvector.mir:402]: There are no tests for non-i1 vectors with non-zero index?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 5]
```diff
@@ -915,6 +921,154 @@ bool RISCVLegalizerInfo::legalizeSplatVector(MachineInstr &MI,
...
+         "Unexpected vector LLT");
+  return LLT::scalable_vector(RISCV::RVVBitsPerBlock /
+                                  VecTy.getElementType().getSizeInBits(),
+                              VecTy.getElementType());
+}
+
+bool RISCVLegalizerInfo::legalizeInsertSubvector(MachineInstr &MI,
+                                                 MachineIRBuilder &MIB) const {
+  assert(MI.getOpcode() == TargetOpcode::G_INSERT_SUBVECTOR);
+
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:935]: Would a GInsertSubvector in GenericMachineInstrs.h with a cast to GInsertSubvector make it easier?

**Comment by tschuett** [comment]
IRTranslator support for G_INSERT_SUBVECTOR is missing.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 5 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:935]: Added

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 6]
```diff
@@ -915,6 +921,152 @@ bool RISCVLegalizerInfo::legalizeSplatVector(MachineInstr &MI,
...
+                              VecTy.getElementType());
+}
+
+bool RISCVLegalizerInfo::legalizeInsertSubvector(MachineInstr &MI,
+                                                 MachineIRBuilder &MIB) const {
+  GInsertSubvector &IS = cast<GInsertSubvector>(MI);
+
+  MachineRegisterInfo &MRI = *MIB.getMRI();
+
+  Register Dst = IS.getOperand(0).getReg();
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:938]: `IS.getReg(0);`

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 7]
```diff
@@ -915,6 +921,152 @@ bool RISCVLegalizerInfo::legalizeSplatVector(MachineInstr &MI,
...
+
+  const RISCVRegisterInfo *TRI = STI.getRegisterInfo();
+  MVT LitTyMVT = getMVTForLLT(LitTy);
+  unsigned SubRegIdx, RemIdx;
+  std::tie(SubRegIdx, RemIdx) =
+      RISCVTargetLowering::decomposeSubvectorInsertExtractToSubRegs(
+          getMVTForLLT(BigTy), LitTyMVT, Idx, TRI);
+
+  RISCVII::VLMUL SubVecLMUL = RISCVTargetLowering::getLMUL(getMVTForLLT(LitTy));
+  bool IsSubVecPartReg = !RISCVVType::decodeVLMUL(SubVecLMUL).second;
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:992]: @topperc 

This code was actually removed in f565b79f9fc27 for SelectionDAG. I wrote this patch before that code existed. I suppose we should take the new code instead?

**Comment by topperc** [comment]
I'm not sure I believe this belongs in the legalizer. This feels like custom instruction selection. I think its similar to the reasons AArch64 created AArch64PostLegalizerLowering. Perhaps we should be doing the same?

**Comment by tschuett** [comment]
What is the difference between hosting this code at:
* the legalizer
* the post legalizer lower combiner
* the instruction selector?

**Comment by topperc** [comment]
> What is the difference between hosting this code at:
> 
> * the legalizer
> 
> * the post legalizer lower combiner
> 
> * the instruction selector?

Emitting a bunch of target specific opcodes from the legalizer makes it very hard to do any optimizations that may be exposed by legalizing the types.

Doing it in instruction selection means we can't reuse tablegen patterns. There are up to 34 different variants of instructions  considering element width and vector width.

Lowering allows us to select target specific generic nodes after after post legalizer optimizations and still use tablegen patterns.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 8]
```diff
@@ -1739,7 +1739,7 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
     }
 
     if (IndexOp.getImm() != 0 &&
-        Src1Ty.getElementCount().getKnownMinValue() % IndexOp.getImm() != 0) {
+        IndexOp.getImm() % Src1Ty.getElementCount().getKnownMinValue() != 0) {
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1742]: This needs to be done in a separate patch with its own unit test.

File: llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp [hunk 9]
```diff
@@ -41,6 +41,184 @@ namespace {
...
+                                            MachineIRBuilder &MIB,
+                                            MachineRegisterInfo &MRI) {
+  LLT MaskTy = getMaskTypeFor(VecTy);
+  return MIB.buildInstr(RISCV::G_VMSET_VL, {MaskTy}, {VL});
+}
+
+/// Gets the two common "VL" operands: an all-ones mask and the vector length.
+/// VecTy is a scalable vector type.
+static std::pair<MachineInstrBuilder, Register>
+buildDefaultVLOps(const DstOp &Dst, MachineIRBuilder &MIB,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:73]: Why does this take a DstOp?

File: llvm/test/CodeGen/RISCV/GlobalISel/postlegalizer-lowering/rvv/insert-subvector.mir [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+    ; RV64-NEXT: PseudoRET implicit $v8
+    %0:_(<vscale x 2 x s8>) = G_IMPLICIT_DEF
+    %1:_(<vscale x 1 x s8>) = G_IMPLICIT_DEF
+    %2:_(<vscale x 2 x s8>) = G_INSERT_SUBVECTOR %0, %1(<vscale x 1 x s8>), 1
+    $v8 = COPY %2(<vscale x 2 x s8>)
+    PseudoRET implicit $v8
+
+...
+
+# i1-element vectors with zero index
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/postlegalizer-lowering/rvv/insert-subvector.mir:296]: Comment seems wrong?

File: llvm/test/CodeGen/RISCV/GlobalISel/postlegalizer-lowering/rvv/insert-subvector.mir [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+    ; RV32-NEXT: PseudoRET implicit $v8
+    ;
+    ; RV64-LABEL: name: insert_subvector_nxv2i16_nxv4i16
+    ; RV64: [[DEF:%[0-9]+]]:_(<vscale x 4 x s16>) = G_IMPLICIT_DEF
+    ; RV64-NEXT: [[VMSET_VL:%[0-9]+]]:_(<vscale x 4 x s1>) = G_VMSET_VL $x0
+    ; RV64-NEXT: [[VSCALE:%[0-9]+]]:_(s64) = G_VSCALE i64 4
+    ; RV64-NEXT: [[VSCALE1:%[0-9]+]]:_(s64) = G_VSCALE i64 1
+    ; RV64-NEXT: [[ADD:%[0-9]+]]:_(s64) = G_ADD [[VSCALE1]], [[VSCALE]]
+    ; RV64-NEXT: [[VSLIDEUP_VL:%[0-9]+]]:_(<vscale x 4 x s16>) = G_VSLIDEUP_VL [[DEF]], [[DEF]], [[VSCALE1]](s64), [[VMSET_VL]](<vscale x 4 x s1>), [[ADD]](s64), 0
+    ; RV64-NEXT: [[BITCAST:%[0-9]+]]:_(<vscale x 4 x s16>) = G_BITCAST [[VSLIDEUP_VL]](<vscale x 4 x s16>)
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/postlegalizer-lowering/rvv/insert-subvector.mir:321]: This bitcast is unnecessary

File: llvm/test/CodeGen/RISCV/GlobalISel/postlegalizer-lowering/rvv/insert-subvector.mir [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+    ; RV32-NEXT: PseudoRET implicit $v8
+    ;
+    ; RV64-LABEL: name: insert_subvector_nxv1i8_nxv2i8
+    ; RV64: [[DEF:%[0-9]+]]:_(<vscale x 2 x s8>) = G_IMPLICIT_DEF
+    ; RV64-NEXT: [[VMSET_VL:%[0-9]+]]:_(<vscale x 2 x s1>) = G_VMSET_VL $x0
+    ; RV64-NEXT: [[VSCALE:%[0-9]+]]:_(s64) = G_VSCALE i64 2
+    ; RV64-NEXT: [[VSCALE1:%[0-9]+]]:_(s64) = G_VSCALE i64 1
+    ; RV64-NEXT: [[ADD:%[0-9]+]]:_(s64) = G_ADD [[VSCALE1]], [[VSCALE]]
+    ; RV64-NEXT: [[VSLIDEUP_VL:%[0-9]+]]:_(<vscale x 2 x s8>) = G_VSLIDEUP_VL [[DEF]], [[DEF]], [[VSCALE1]](s64), [[VMSET_VL]](<vscale x 2 x s1>), [[ADD]](s64), 0
+    ; RV64-NEXT: [[BITCAST:%[0-9]+]]:_(<vscale x 2 x s8>) = G_BITCAST [[VSLIDEUP_VL]](<vscale x 2 x s8>)
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/postlegalizer-lowering/rvv/insert-subvector.mir:285]: This bitcast is unnecessary

File: llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp (hunk 9 above)
Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:73]: Should probably be `LLT VecTy`.

File: llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp [hunk 13]
```diff
@@ -41,6 +41,184 @@ namespace {
...
+  LLT InterLitTy = BigTy;
+  Register AlignedExtract = Src1;
+  unsigned AlignedIdx = Idx - RemIdx;
+  if (TypeSize::isKnownGT(BigTy.getSizeInBits(),
+                          getLMUL1Ty(BigTy).getSizeInBits())) {
+    InterLitTy = getLMUL1Ty(BigTy);
+    // Extract a subvector equal to the nearest full vector register type. This
+    // should resolve to a G_EXTRACT on a subreg.
+    AlignedExtract =
+        MIB.buildExtractSubvector(InterLitTy, BigVec, AlignedIdx).getReg(0);
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: Nit: You haven't legalized G_EXTRACT_SUBVECTOR yet. You build an illegal instruction after the legalizer?

Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: Chicken and egg problem?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: I brought this up to @topperc this morning. I thought it was a little odd that this is how the post legalizer lowering pass works: you can introduce non-legal instructions and it wont complain. But it will complain when you get to isel because we won't be able to select it.

We can either take it like this and let it fail in isel, or I can lump the G_EXTRACT_SUBVECTOR legalization into this patch. Either works for me. Figured I'd keep the patches small for easier review.

Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: I would leave it as is.

Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: The odd thing about this pass is that the combines are free-standing functions. For the normal combiner, they are member functions and we get access to LegalizerInfo.

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: Can't you make an extract_subvector legalizer patch without lowering?

Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: Isn't that the chicken and egg problem. You make G_EXTRACT_SUBVECTOR legal, but you can't select it.

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: > Isn't that the chicken and egg problem. You make G_EXTRACT_SUBVECTOR legal, but you can't select it.

Isn't that often how GISel patches are staged though?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVPostLegalizerLowering.cpp:178]: If this code was in the legalizer, the legalization test case would fail until G_EXTRACT_SUBVECTOR legalization support is added. Doing the lowering here does not error out due to that.

**Comment by tschuett** [comment]
I stated before: the odd thing about the lower passes: they are free-standing functions. There is no CombinerHelper and no access to LegalizerInfo.
In the combiner, we have the pattern:
```
if (!isLegal(op))
  return false;
build(Op);
```

The lowering combines cannot really fail. How about:
```
assert(isLegal(Op) && "unexpected"):
build(Op);
```

**Comment by arsenm** [comment]
> I stated before: the odd thing about the lower passes: they are free-standing functions. There is no CombinerHelper and no access to LegalizerInfo. In the combiner, we have the pattern:
> 

I don't understand the lowering passes. They seem like they're just a hack and I don't understand why they are needed.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 14]
```diff
@@ -581,6 +581,12 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
 
   SplatActions.clampScalar(1, sXLen, sXLen);
 
+  getActionDefinitionsBuilder(G_INSERT_SUBVECTOR)
+      .legalIf(all(typeIsLegalBoolVec(0, BoolVecTys, ST),
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:585]: It should be `customIf` again.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 15]
```diff
@@ -915,6 +928,150 @@ bool RISCVLegalizerInfo::legalizeSplatVector(MachineInstr &MI,
   return true;
 }
 
+/// Lowers G_INSERT_SUBVECTOR. We know we can lower it here since the legalizer
+/// marked it as legal.
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:932]: It is not marked legal. You marked it as custom.

**Comment by michaelmaitland** [comment]
Thanks @tschuett, I forgot to commit the changes you mention. And thanks also for updating the title. I have to post my patch for legalize extract subvector, because this patch depends on that one.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 14 above)
Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:585]: It is still legalIf.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:585]: I don't know how my change didn't get added once again. Im sorry.

**Comment by michaelmaitland** [comment]
I've updated this patch based on the discussions we had during the legalization of EXTRACT_SUBVECTOR. Now ready for review :)

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 16]
```diff
@@ -1054,6 +1058,142 @@ bool RISCVLegalizerInfo::legalizeExtractSubvector(MachineInstr &MI,
...
+      return Helper.bitcast(
+          IS, 0,
+          LLT::vector(BigTy.getElementCount().divideCoefficientBy(8), 8));
+
+    // We can't slide this mask vector up indexed by its i1 elements.
+    // This poses a problem when we wish to insert a scalable vector which
+    // can't be re-expressed as a larger type. Just choose the slow path and
+    // extend to a larger type, then truncate back down.
+    BigTy = BigTy.changeElementType(LLT::scalar(8));
+    LitTy = LitTy.changeElementType(LLT::scalar(8));
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1097]: Use a new variable here instead of reassigning

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 17]
```diff
@@ -1054,6 +1058,142 @@ bool RISCVLegalizerInfo::legalizeExtractSubvector(MachineInstr &MI,
...
+    auto BigZExt = MIB.buildZExt(BigTy, BigVec);
+    auto LitZExt = MIB.buildZExt(LitTy, LitVec);
+    auto Insert = MIB.buildInsertSubvector(BigTy, BigZExt, LitZExt, Idx);
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1100]: Could make this widenscalar?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 18]
```diff
@@ -1054,6 +1058,142 @@ bool RISCVLegalizerInfo::legalizeExtractSubvector(MachineInstr &MI,
...
+    return true;
+
+  // We don't have the ability to slide mask vectors up indexed by their i1
+  // elements; the smallest we can do is i8. Often we are able to bitcast to
+  // equivalent i8 vectors. Otherwise, we can must zeroextend to equivalent i8
+  // vectors and truncate down after the insert.
+  if (LitTy.getElementType() == LLT::scalar(1)) {
+    auto BigTyMinElts = BigTy.getElementCount().getKnownMinValue();
+    auto LitTyMinElts = LitTy.getElementCount().getKnownMinValue();
+    if (BigTyMinElts >= 8 && LitTyMinElts >= 8)
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1087]: Could do this in the legalization rule instead of in the custom logic

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1087]: Are you sure?

Above is this code is

```
if (Idx == 0 ||
      MRI.getVRegDef(BigVec)->getOpcode() == TargetOpcode::G_IMPLICIT_DEF)
```

This means that the legalization rule needs to check the immediate value to make sure its not zero, which I couldn't figure out how to do, and we need to check to see if the BigVec is not a G_IMPLICIT_DEF, which I also don't know how to do as part of the legalization rule.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1087]: Maybe we can do the under check in a combine, but I still don’t think it’s possible to check that the index is not zero is it?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1087]: Also, if we promote the INSERT_SUBVECTOR just to do the combine later, I am concerned that we will be left with an INSERT_SUBVECTOR that has a larger type than needed.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 17 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1100]: But we're widening a vector?

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1100]: Widening a vector implies changing the element count. Widening scalar means changing the element size even if its part of a vector.

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1100]:
```
  /// The operation should be implemented in terms of a wider scalar             
  /// base-type. For example a <2 x s8> add could be implemented as a <2         
  /// x s32> add (ignoring the high bits).                                       
  WidenScalar,                                                                   
                
```

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1100]: I see, thanks for the clarification.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 19]
```diff
@@ -1054,6 +1058,134 @@ bool RISCVLegalizerInfo::legalizeExtractSubvector(MachineInstr &MI,
...
+      STI.expandVScale(LitTy.getSizeInBits())
+          .isKnownMultipleOf(STI.expandVScale(VecRegSize));
+
+  // If the Idx has been completely eliminated and this subvector's size is a
+  // vector register or a multiple thereof, or the surrounding elements are
+  // undef, then this is a subvector insert which naturally aligns to a vector
+  // register. These can easily be handled using subregister manipulation.
+  if (RemIdx == 0 &&
+      (ExactlyVecRegSized ||
+       MRI.getVRegDef(BigVec)->getOpcode() == TargetOpcode::G_IMPLICIT_DEF))
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1119]: Repeated this check above

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 20]
```diff
@@ -1054,6 +1058,134 @@ bool RISCVLegalizerInfo::legalizeExtractSubvector(MachineInstr &MI,
...
+        MIB.buildInstr(RISCV::G_VSLIDEUP_VL, {InterLitTy},
+                       {AlignedExtract, Insert, SlideupAmt, Mask, VL, Policy});
+  }
+
+  // If required, insert this subvector back into the correct vector register.
+  // This should resolve to an INSERT_SUBREG instruction.
+  if (TypeSize::isKnownGT(BigTy.getSizeInBits(), InterLitTy.getSizeInBits()))
+    MIB.buildInsertSubvector(Dst, BigVec, LitVec, AlignedIdx);
+  else
+    Inserted->getOperand(0).setReg(Dst);
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:1183]: DIdn't call the observer before modifying instruction in place. Can you build the instruction above with the correct register?

**Review by arsenm — APPROVED** [review]
